rest_framework/schemas.py

```diff
@@ -246,7 +246,10 @@ def get_allowed_methods(self, callback):
         Return a list of the valid HTTP methods for this endpoint.
         """
         if hasattr(callback, 'actions'):
-            return [method.upper() for method in callback.actions.keys()]
+            return [
+                method.upper() for method in
+                set(callback.actions.keys()).intersection(set(callback.cls().http_method_names))
```

Comment:
There's too much going on inline here.
I'd suggest:

* Pull each of those two sets out into variable assignments.
* Use the `&` operator, rather then `.intersection(...)`